src/FreeAge/server/game.cpp

```diff
@@ -1349,21 +1401,137 @@ bool Game::SimulateMeleeAttack(u32 /*unitId*/, ServerUnit* unit, u32 targetId, S
+    newUnit->SetMapCoord(building->GetBaseTile());
+  }
+  
+  // Send messages to clients that see the new unit
+  QByteArray addObjectMsg = CreateAddObjectMessage(newUnitId, newUnit);
+  for (auto& player : *playersInGame) {
+    accumulatedMessages[player->index] += addObjectMsg;
+  }
+
+  if (!foundFreeSpace) {
+    // TODO: what happens when the capacity is full ?
+    building->GarrisonUnit(newUnit);
+    newUnit->SetGarrisonedInsideObject(building);
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(newUnitId, buildingId);
+    for (auto& player : *playersInGame) {
+      accumulatedMessages[player->index] += unitGarrisonMessage;
+    }
+  }
+  
+  GetPlayerStats(newUnit->GetPlayerIndex())->UnitAdded(unitInProduction);
+}
+
+void Game::GarrisonUnit(u32 unitId, ServerUnit* unit, u32 targetObjectId, ServerObject* targetObject, bool enter, bool* unitMovementChanged) {
+  // NOTE: Could be simplified if Objects stored there ID. #ids
+  *unitMovementChanged = false;
+  if (enter) {
+    // TODO: test with target object being a unit
+
+    unit->StopMovement();
+    unit->RemoveTarget();
+
+    // Check if the object is a building and has <= 20% hp left
+    if (targetObject->GetGarrisonedUnitsCount() > 0 && targetObject->isBuilding() &&
+        AsBuilding(targetObject)->GetHPInternalFloat() <= .2f * GetBuildingMaxHP(AsBuilding(targetObject)->GetType())) {
+      LOG(WARNING) << "Unit cannot garrison building beacause of low health";
+      *unitMovementChanged = true;
+      return;
+    }
+    
+    // TODO: replace with the garrison capacity form the unit type stats #stats
+    bool isTownCenter = targetObject->GetObjectType() == ObjectType::Building && AsBuilding(targetObject)->GetType() == BuildingType::TownCenter;
+    int capacity = isTownCenter ? 15 : 0;
+    if (targetObject->GetGarrisonedUnitsCount() >= capacity) {
+      LOG(WARNING) << "Unit cannot garrison building beacause of filled capacity";
+      *unitMovementChanged = true;
+      return;
+    }
+
+    targetObject->GarrisonUnit(unit);
+    unit->SetGarrisonedInsideObject(targetObject);
+
+    QByteArray unitGarrisonMessage = CreateUnitGarrisonMessage(unitId, targetObjectId);
+    QByteArray unitMoveMessage = CreateUnitMovementMessage(unitId,
+        unit->GetMapCoord(),
+        unit->GetMoveSpeed() * unit->GetMovementDirection(),
+        unit->GetCurrentAction());
+    for (auto& player : *playersInGame) {
+      accumulatedMessages[player->index] += unitMoveMessage;
+      accumulatedMessages[player->index] += unitGarrisonMessage;
+    }
+  } else {
+    if (!targetObject->isBuilding()) {
+      // TODO: find an empty spot around the target unit and ungarrison their
+      LOG(ERROR) << "Ungarrison from unit not implemented yet";
+      return;
+    }
+    ServerBuilding* targetBuilding = AsBuilding(targetObject);
+    if (QPointF freeSpace; FindFreeSpaceAroundBuilding(targetBuilding, unit, freeSpace)) {
```

Comment:
I am surprised that this compiles, can't remember seeing multiple statements within an `if()` before :) I would thus suggest to move the `QPointF freeSpace;` out of the condition to keep this in line with what is usually done.

Reply:
This is a new C++17 thing, I believe is inspired from Go. It basically makes freeSpace only usable inside the if. Are you a fun of new things or should I just remove it?

_[Selection statements with initializer](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0305r1.html)_

Reply:
Hmm okay. Since I was not used to it, it looked like an error to me, but the problem is on my side then: I should get used to it :)

Reply:
Second time I used it... but we must evolve!

Bonus: Leaving it in, we also cause other people reading to evolve also.